<img style="float: left; margin: 30px 15px 15px 15px;" src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTFzQj91sOlkeDFkg5HDbjtR4QJYmLXkfMNig&usqp=CAU" width="400" height="500" /> 
    
    
### <font color='navy'> Simulación matemática. 

**Nombres:** 
- Samuel Rodríguez Partida | ic727594
- Gustavo de Anda Madrigal | if742126
    
**Fecha:** 14 de noviembre del 2022.

    
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github**: https://github.com/Samuelrodriguezp

# Tarea 8

## Enunciado de tarea (Transcibirlo)

## Tarea.
> - Consultar tasas de interés reales en algún banco y proyectar un ahorro mensual para que al terminar su carrera tengan $\$85000$ MXN en su cuenta.
> - Hacer esto en un nuevo archivo, llamarlo Gi_T7ApellidoNombre.ipynb y subirlo a canvas. Donde $i=1,2$ es el número del grupo al que pertenece 
> - Plazo: 

## CÓDIGO DE SOLUCIÓN ALUMNO 1

Se tomó como referencia los rendimientos de certificados de la tesorería (CETES) a un plazo de un año, siendo *10.87* el rendimiento

In [1]:
#Instalamos nuestros paquetes y librerías

import pandas as pd
from datetime import date
import numpy as np
import numpy_financial as npf

In [2]:
#Datos para la cuenta
tinteres = 0.1087
#Planeamos una fecha de término de la carrera en junio del 2025
año = 2.5
abono_año = 12
#Iniciamos en nuestra cuenta con 5000 pesos
inicial = 5000
meta = 85000
fecha_inicio = (date(2022,11,14))

In [3]:
#calcular abonos mensuales
abono = -npf.pmt(tinteres/abono_año,año*abono_año,(meta-inicial))
abono

3057.3765334219124

#### Debemos hacer abonos de 3,057 pesos

In [4]:
#Empezamos a generar nuestro dataframe y los valores necesarios
rng = pd.date_range(fecha_inicio, periods=año*abono_año, freq="M")
rng.name = 'Fechas del abono'

#Columnas de abonos
df = pd.DataFrame(index = rng, columns=['Abono','Balance'],dtype='float')
#Fechas como índices
df.reset_index(inplace=True)

df.index+=1
df.index.name = 'Periodo'
df.head()

Fechas del abono  Abono  Balance
Periodo                                 
1             2022-11-30    NaN      NaN
2             2022-12-31    NaN      NaN
3             2023-01-31    NaN      NaN
4             2023-02-28    NaN      NaN
5             2023-03-31    NaN      NaN

In [5]:
#Definir abonos
df['Abono'] = abono
#Vector de los periodos
index_vector = np.arange(1,len(rng)+1)
#Balance de la cuenta objetivo
df['Balance'] = inicial * (1+tinteres/abono_año)**index_vector + \
                abono *                       (((1+tinteres/abono_año)**index_vector -1)/(tinteres/abono_año))


#Redondeo de cifras para mejor formato
df = df.round(2)
df

Fechas del abono    Abono    Balance
Periodo                                     
1             2022-11-30  3057.38    8102.67
2             2022-12-31  3057.38   11233.44
3             2023-01-31  3057.38   14392.57
4             2023-02-28  3057.38   17580.32
5             2023-03-31  3057.38   20796.95
6             2023-04-30  3057.38   24042.71
7             2023-05-31  3057.38   27317.87
8             2023-06-30  3057.38   30622.71
9             2023-07-31  3057.38   33957.47
10            2023-08-31  3057.38   37322.45
11            2023-09-30  3057.38   40717.90
12            2023-10-31  3057.38   44144.12
13            2023-11-30  3057.38   47601.36
14            2023-12-31  3057.38   51089.93
15            2024-01-31  3057.38   54610.10
16            2024-02-29  3057.38   58162.15
17            2024-03-31  3057.38   61746.38
18            2024-04-30  3057.38   65363.07
19            2024-05-31  3057.38   69012.53
20            2024-06-30  3057.38   72695.05
21            2024-07-31  3057.38   76410.92
22            2024-08-31  3057.38   80160.45
23            2024-09-30  3057.38   83943.95
24            2024-10-31  3057.38   87761.72
25            2024-11-30  3057.38   91614.07
26            2024-12-31  3057.38   95501.31
27            2025-01-31  3057.38   99423.77
28            2025-02-28  3057.38  103381.76
29            2025-03-31  3057.38  107375.61
30            2025-04-30  3057.38  111405.63

### Respuesta planteada en el ejercicio usando (markdown)
 La respuesta es que teniendo una cuenta con 5,000 pesos y buscando un saldo objetivo de 85,000 pesos, tendré que invertir en cetes un año al 10.87% (suponiendo que el rendimiento se mantenga siempre igual) todos los meses la cantidad de 3,057.38 pesos para que en noviembre del 2024 consiga mi objetivo.

## CÓDIGO DE SOLUCIÓN ALUMNO 2

In [6]:
import pandas as pd
from datetime import date
import numpy as np
import matplotlib.pyplot as plt
import numpy_financial as npf
%matplotlib inline

In [7]:
tasa_interes = 0.0775
years = 3.5
Abono_year = 12
inicial = 8000
meta = 85000
fecha_inicio = (date(2022,11,14))

La tasa de interés del 7.75% se extrajo del Grupo Bursátil Mexicano (GBM). 

In [8]:
# Calcular abonos mensuales para llegar a la meta de ahorro en el tiempo requerido.
Abono = -npf.pmt(tasa_interes/Abono_year,years*Abono_year,(meta-inicial))
Abono

2099.0840461570538

### Como podemos ver, al calcular los abonos, nuestro código nos dice que debemos de abonar 2099.08 pesos de manera mensual para llegar a nuestra meta.

In [9]:
# Fechas de la tabla 
rng = pd.date_range(fecha_inicio, periods=years*Abono_year, freq='M')
rng.name = 'Fecha del abono'
# Columnas de abonos y extracto de cuenta en las fechas respectivas
df = pd.DataFrame(index = rng,columns=['Abono','Balance'],dtype='float')
# Indice con los periodos
df.reset_index(inplace=True)

df.index+=1
df.index.name='Periodo'
df.head()

Fecha del abono  Abono  Balance
Periodo                                
1            2022-11-30    NaN      NaN
2            2022-12-31    NaN      NaN
3            2023-01-31    NaN      NaN
4            2023-02-28    NaN      NaN
5            2023-03-31    NaN      NaN

In [10]:
npf.ppmt(tasa_interes/Abono_year, 1, years * Abono_year, Abono)

-43.66619258498984

In [11]:
# Los abonos son iguales
df['Abono'] = Abono
# Vector indizador de los periodos de los abonos
index_vector = np.arange(1,len(rng)+1)
# Extracto de la cuenta mes a mes
df["Balance"] = inicial * (1 + tasa_interes/Abono_year)**index_vector + \
                Abono * (((1 + tasa_interes/Abono_year)**index_vector - 1)/(tasa_interes/Abono_year))

df.head()

Fecha del abono        Abono       Balance
Periodo                                           
1            2022-11-30  2099.084046  10150.750713
2            2022-12-31  2099.084046  12315.391691
3            2023-01-31  2099.084046  14494.012641
4            2023-02-28  2099.084046  16686.703853
5            2023-03-31  2099.084046  18893.556194

In [12]:
# Redondear con dos cifras decimales y mostrar DataFrame
df = df.round(2)
df

Fecha del abono    Abono    Balance
Periodo                                    
1            2022-11-30  2099.08   10150.75
2            2022-12-31  2099.08   12315.39
3            2023-01-31  2099.08   14494.01
4            2023-02-28  2099.08   16686.70
5            2023-03-31  2099.08   18893.56
6            2023-04-30  2099.08   21114.66
7            2023-05-31  2099.08   23350.11
8            2023-06-30  2099.08   25600.00
9            2023-07-31  2099.08   27864.41
10           2023-08-31  2099.08   30143.46
11           2023-09-30  2099.08   32437.22
12           2023-10-31  2099.08   34745.79
13           2023-11-30  2099.08   37069.28
14           2023-12-31  2099.08   39407.77
15           2024-01-31  2099.08   41761.36
16           2024-02-29  2099.08   44130.15
17           2024-03-31  2099.08   46514.24
18           2024-04-30  2099.08   48913.73
19           2024-05-31  2099.08   51328.72
20           2024-06-30  2099.08   53759.30
21           2024-07-31  2099.08   56205.58
22           2024-08-31  2099.08   58667.66
23           2024-09-30  2099.08   61145.63
24           2024-10-31  2099.08   63639.62
25           2024-11-30  2099.08   66149.71
26           2024-12-31  2099.08   68676.01
27           2025-01-31  2099.08   71218.63
28           2025-02-28  2099.08   73777.66
29           2025-03-31  2099.08   76353.23
30           2025-04-30  2099.08   78945.43
31           2025-05-31  2099.08   81554.37
32           2025-06-30  2099.08   84180.16
33           2025-07-31  2099.08   86822.90
34           2025-08-31  2099.08   89482.72
35           2025-09-30  2099.08   92159.71
36           2025-10-31  2099.08   94853.99
37           2025-11-30  2099.08   97565.68
38           2025-12-31  2099.08  100294.87
39           2026-01-31  2099.08  103041.69
40           2026-02-28  2099.08  105806.26
41           2026-03-31  2099.08  108588.67
42           2026-04-30  2099.08  111389.06

### Al final de la carrera, por estar abonando de manera mensual  2099.08 pesos, saldríamos con el valor de una cuenta de ahorro de 111,389.06 pesos 

## Bibliografía
- GBM. (2022, 3 octubre). GBM+ | Smart Cash. GBM+. https://gbm.com/plus/smart-cash/
- Cetesdirecto (14,10,2022). [En línea]. https://www.cetesdirecto.com/sites/portal/inicio